test/segment-loader.test.js

```diff
@@ -1865,6 +1931,213 @@ QUnit.module('SegmentLoader', function(hooks) {
+          appends.push(config);
+          origAppendToSourceBuffer(config);
+        };
+        sourceUpdater.audioTimestampOffset = (offset) => {
+          if (!offset) {
+            return audioTimestampOffsets.length ?
+              audioTimestampOffsets[audioTimestampOffsets.length - 1] : -1;
+          }
+          audioTimestampOffsets.push(offset);
+          origAudioTimestampOffset(offset);
+        };
+        sourceUpdater.videoTimestampOffset = (offset) => {
+          if (!offset) {
+            return videoTimestampOffsets.length ?
+              videoTimestampOffsets[videoTimestampOffsets.length - 1] : -1;
+          }
+          videoTimestampOffsets.push(offset);
+          origVideoTimestampOffset(offset);
+        };
+        loader.transmuxer_.postMessage = (message) => {
+          if (message.action === 'setTimestampOffset') {
+            transmuxerTimestampOffsets.push(message.timestampOffset);
+          }
+          origTransmuxerPostMessage(message);
+        };
+
+        loader.playlist(playlistWithDuration(20));
+        loader.load();
+        this.clock.tick(1);
+        standardXHRResponse(this.requests.shift(), zeroLengthSegment());
+        return new Promise((resolve, reject) => {
+          loader.one('appended', resolve);
+          loader.one('error', reject);
+        });
+      }).then(() => {
+        assert.equal(appends.length, 0, 'zero appends');
+        assert.equal(
+          audioTimestampOffsets.length,
+          0,
+          'zero audio source buffer timestamp offsets'
+        );
+        assert.equal(
+          videoTimestampOffsets.length,
+          0,
+          'zero video source buffer timestamp offsets'
+        );
+        // unlike the source buffer, which won't have data appended yet, the transmuxer
+        // timestamp offset should be updated since there may be ID3 data or metadata
+        assert.equal(
+          transmuxerTimestampOffsets.length,
+          1,
+          'one transmuxer timestamp offset'
+        );
+
+        this.clock.tick(1);
+        standardXHRResponse(this.requests.shift(), muxedSegment());
+        return new Promise((resolve, reject) => {
+          loader.one('appended', resolve);
+          loader.one('error', reject);
+        });
+      }).then(() => {
+        this.clock.tick(1);
+
+        assert.equal(appends.length, 2, 'two appends');
+        assert.equal(
+          audioTimestampOffsets.length,
+          1,
+          'one audio source buffer timestamp offset'
+        );
+        assert.equal(
+          videoTimestampOffsets.length,
+          1,
+          'one video source buffer timestamp offset'
+        );
+        assert.equal(
+          transmuxerTimestampOffsets.length,
+          2,
+          'another transmuxer timestamp offset'
+        );
+      });
+    });
+
+    QUnit.test('sets timestamp offset on timeline changes but not if segment start is early', function(assert) {
```

Comment:
this is so much less readable than https://github.com/videojs/http-streaming/pull/502/files#diff-4efdae9dc0d9508ec02d8c6707ae5df8R445 but idk if there's really anything we can about that.

Reply:
Maybe it can be annotated with some comments about what's happening?